### PR TITLE
Friends integrity with 2 rows.

### PR DESCRIPTION
**Pull request by Meltrust**
In this milestone I:

-    Reviewed and created models for the project requirements.
-    Added helper methods for the mirrored friendships in the 2-row context.
-    Updated models, controllers, and helpers.
-    Followed the git-flow rules.
-    Encapsulated the code in classes and objects.
-    Delivered the project with an organized folder structure.
-    Updated gems
-    Used the simplest solutions as possible, over the most complex ones. Readability above all was always the goal.
-    Wrote a descriptive PR.

Notice:  After the last milestone, I forgot to merge.  After realizing this, I had to merge so some of my commits are on the last merge.  In order to preserve integrity, I had to continue in this same branch instead of creating another one.  Won't happen again.  Thanks.

Thanks for the review!

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -10,11 +10,11 @@ class User < ApplicationRecord
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
   has_many :friendships
-  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
+  # has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
```
Comment (rindrajosia) [inline app/models/user.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/models/user.rb [hunk 2]
```diff
@@ -10,11 +10,11 @@ class User < ApplicationRecord
...
 
   def friends
     friends_array = friendships.map { |friendship| friendship.friend if friendship.confirmed }
-    friends_array += inverse_friendships.map { |friendship| friendship.user if friendship.confirmed }
+    # friends_array += inverse_friendships.map { |friendship| friendship.user if friendship.confirmed }
```
Comment (rindrajosia) [inline app/models/user.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/models/user.rb [hunk 3]
```diff
@@ -25,17 +25,20 @@ def pending_friends
 
   # Users who have requested to be friends
   def friend_requests
-    inverse_friendships.map { |friendship| friendship.user unless friendship.confirmed }.compact
+    # inverse_friendships.map { |friendship| friendship.user unless friendship.confirmed }.compact
```
Comment (rindrajosia) [inline app/models/user.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/models/user.rb [hunk 4]
```diff
@@ -25,17 +25,20 @@ def pending_friends
...
+    Friendship.all.where(friend_id: id, confirmed: false)
   end
 
   def confirm_friend(user)
-    friendship = inverse_friendships.find { |my_friendship| my_friendship.user == user }
+    # friendship = inverse_friendships.find { |my_friendship| my_friendship.user == user }
```
Comment (rindrajosia) [inline app/models/user.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/models/user.rb [hunk 5]
```diff
@@ -25,17 +25,20 @@ def pending_friends
...
+    friendship = friend_requests.find { |f| f.user == user }
     friendship.confirmed = true
     friendship.save
   end
 
   def reject_friend(user)
-    friendship = inverse_friendships.find { |my_friendship| my_friendship.user == user }
+    # friendship = inverse_friendships.find { |my_friendship| my_friendship.user == user }
```
Comment (rindrajosia) [inline app/models/user.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/models/friendship.rb [hunk 6]
```diff
@@ -1,4 +1,41 @@
...
+
   belongs_to :user
   belongs_to :friend, class_name: 'User'
+
+  def set_defaults
+    self.confirmed = false if confirmed.nil?
+  end
+
+  def create_inverse
+    # Friendship.create(user: user, friend: friend, confirmed: nil)
```
Comment (rindrajosia) [inline app/models/friendship.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -32,14 +32,16 @@ def accept_friendship(friendship)
...
+    # link_to('Accept?', user_friendship_path(user_id: friendship.id), method: :put, class: 'profile-link')
+    link_to('Accept?', user_friendship_path(user_id: friendship.user.id), method: :put, class: 'profile-link')
   end
 
   def reject_friendship(friendship)
     show_user = User.find_by(id: params[:id])
     (return unless current_user == show_user)
 
-    link_to('Reject?', user_friendship_path(user_id: friendship.id), method: :delete, class: 'profile-link')
+    # link_to('Reject?', user_friendship_path(user_id: friendship.id), method: :delete, class: 'profile-link')
```
Comment (rindrajosia) [inline app/helpers/application_helper.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -32,14 +32,16 @@ def accept_friendship(friendship)
     show_user = User.find_by(id: params[:id])
     (return unless current_user == show_user)
 
-    link_to('Accept?', user_friendship_path(user_id: friendship.id), method: :put, class: 'profile-link')
+    # link_to('Accept?', user_friendship_path(user_id: friendship.id), method: :put, class: 'profile-link')
```
Comment (rindrajosia) [inline app/helpers/application_helper.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

File: app/controllers/friendships_controller.rb [hunk 9]
```diff
@@ -10,7 +10,8 @@ def create
   end
 
   def update
-    friend = User.find(params[:user_id])
+    # friend = User.find(params[:user_id])
```
Comment (rindrajosia) [inline app/controllers/friendships_controller.rb]: Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- User can send friendship request :+1: 

To improve :factory_worker: [REQUIRED]:

- [x] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)) and when friendship is confirmed, an additional row with reversed user-friend pair is created
```
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [ ] thanks to this, we can find user friends only by user_id in friendships table
```
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

- [ ] User#pending_friends can be defined as association with `SQL query instead of processing objects in application memory` (as the database was created to store and query data it is always more efficient than ruby)
```
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend


```

- [ ] User#friend_requests can be defined as association with `SQL query instead of processing objects in application memory` (as the database was created to store and query data it is always more efficient than ruby)
```
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships

```

- [ ] in order to get all posts written by user on their friends, students can use simple method in User model (in order to **NOT** do it with Enumerables)
```
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

- [ ] Inline comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history.

- [ ] Please add more unit tests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://www.theodinproject.com/courses/ruby-on-rails/lessons/final-project))

Nice to have [OPTIONAL] :

- [ ] It will be nice if you put all logic in your view inside helper.
- [ ] Please read about n+1 and make sure there are no n + 1 queries in index views [Recommended article.](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5)

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Review by oloomoses — APPROVED** [review]
Hello @Meltrust,
**Well done on implementing the changes suggested by the @rindrajosia**
### Highlights
- An additional row is created on inverse friendship :heavy_check_mark:
- Unit and integration tests :heavy_check_mark:
- Descriptive pr :heavy_check_mark:
- Linters and rubocop configuration :heavy_check_mark:

**This milestone meets all the requirements and therefore:**
## Approved :tada: :muscle: :clap:
Congratulations and best wishes on the next milestone. 

#### Happy coding
_slack: Oloo Moses_

**Pull request merged**